File: rust/src/sqlparser/parser.rs

use super::ast;
use super::tokens;
pub trait Parser{
    //Trait/Interface for the parser. Given the 
    fn parse(&self, query : &str) -> (ast::Query, ast::ErrorCode);
}
pub struct SQLParser {
    
    pub sql : String,
    pub step: tokens::Step,
    pub query : ast::Query,
    pub error : ast::ErrorCode,
}
impl SQLParser {
    
    pub fn new() ->Self {
        let p = SQLParser{
            sql : String::from(""),
            step: tokens::Step::stepType,
            query: ast::Query::new(), 
            error: ast::ErrorCode::None,
            
        };
        p
    } 
    
}
impl Parser for SQLParser {
    fn parse(&self, query: &str) -> (ast::Query, ast:: ErrorCode) {
        let tokenizer = tokens::Tokenizer::new();
        let mut q = ast::Query::new();
        let tokens = tokenizer.get_tokens(String::from(query));
        let mut updateNextField : &str = ""; 
        if !tokens.is_ok() {
            return (q, ast::ErrorCode::TokenizationError(String::from("Unable to parse input query")))
        } 
        let  mut step: tokens::Step = tokens::Step::stepType;
        for token in tokens.unwrap_or(vec![]).iter().peekable(){
            match step {
                tokens::Step::stepType => {
                    if String::from("SELECT").eq_ignore_ascii_case(&token[..]) {
                        // we received a select as the first state . WE change the query type to select and proceed to the next state. 
                        q.query_type = ast::QueryType::SELECT;
                        step = tokens::Step::selectField;
                    }
                    else if String::from("INSERT INTO").eq_ignore_ascii_case(&token[..]) {
                        // here we validate whether the next token is indeed from 
                        q.query_type = ast::QueryType::INSERT;
                        step = tokens::Step::insertTable;
                    }
                    else if String::from("DELETE FROM").eq_ignore_ascii_case(&token[..]) {
                        q.query_type = ast::QueryType::DELETE;
                        step = tokens::Step::deleteFromTable;
                    }
                    else if String::from("UPDATE").eq_ignore_ascii_case(&token[..]) {
                        q.query_type = ast::QueryType::UPDATE;
                        step = tokens::Step::updateTable;
                    
                        //TODO: add metadata for a select query to the query struct

                    }
                    else {
                        return (q,ast::ErrorCode::ParseError(String::from("Invalid query type")))

                    
                    }
                    
                }
                tokens::Step::selectField => {
                    if !tokenizer.is_identifier_or_asterix(token){
                        return (q,ast::ErrorCode::ParseError
                        (String::from("at SELECT: expected field to select")));
                    }
                    q.fields.push(String::from(&token[..]));
                    step = tokens::Step::validateSelectFromOrComma;
                }
                tokens::Step::validateSelectFromOrComma => {
                    if String::from(",").eq_ignore_ascii_case(&token[..]){
                        step = tokens::Step::selectField;
                    }
                    else if String::from("FROM").eq_ignore_ascii_case(&token[..]) {
                        step = tokens::Step::selectFromTable;
                        continue;
                    
                    }
                    else {
                        //TODO: return proper error
                        return (q, ast::ErrorCode::ParseError(
                            String::from("at SELECT expected FROM or a comma after the field")));
                    }
                
                }
                tokens::Step::selectFromTable => {
                    if token.len() == 0{
                        //TODO: return proper error
                        return (q, ast::ErrorCode::ParseError(
                            String::from("at SELECT: expected quoted table name")
                        )
                        );
                    }
                    q.table = String::from(token);
                    step = tokens::Step::stepWhere;
                }
                tokens::Step::insertTable => {
                    if token.len() == 0 {
                        //TODO: throw  proper error
                        return (q, ast::ErrorCode::ParseError(
                            String::from("at INSERT INTO: expected quoted table name")
                        ));
                    }
                    q.table = String::from(token);
                    step = tokens::Step::insertFieldsOpeningParens;
                }
                tokens::Step::deleteFromTable =>{
                    if token.len() == 0 {
                        //TODO: throw  proper error
                        return (q, ast::ErrorCode::ParseError(
                            String::from("at DELETE FROM: expected quoted table name")
                        ));
                    }
                    q.table = String::from(token);
                    step = tokens::Step::stepWhere;
                }
                tokens::Step::updateTable => {
                    if token.len() == 0 {
                     //TODO: throw  proper error
                     return (q, ast::ErrorCode::ParseError(
                         String::from("at UPDATE: expected quoted table name")
                     )
                    );
                    }
                    q.table = String::from(token);
                    step = tokens::Step::updateSet;
                }
                tokens::Step::updateSet => {
                    if ! String::from("SET").eq_ignore_ascii_case(&token[..]) {
                        //TODO: throw proper error code
                        return (q, ast::ErrorCode::ParseError(
                            String::from("at UPDATE: expected 'SET'")
                        ));
                    }
                    step = tokens::Step::updateField;
                }
                tokens::Step::updateField =>{
                    if ! tokenizer.is_identifier(token){
                        //TODO: throw proper error code
                        return (q, ast::ErrorCode::ParseError(
                            String::from("At UPDATE: expected atleast one field to update")
                        ));
                    }
                    updateNextField = &token[..];
                    step = tokens::Step::updateEquals;
                }
                tokens::Step::updateEquals => {
                    if ! String::from("=").eq_ignore_ascii_case(&token[..]){
                        //TODO: throw proper error code
                        return (q, ast::ErrorCode::ParseError(
                            String::from("at UPDATE: expected '='")
                        )
                        );
                    }
                    step = tokens::Step::updateValue;
                }
                
                tokens::Step::updateValue =>{
                    if token.len() == 0 {
                        //TODO: throw proper error code
                        return (q, ast::ErrorCode::ParseError(
                            String::from("at UPDATE: expected quoted value")
                        )
                        );
                    }
                    //TODO: add update fields 
                    q.update_fields.insert(String::from(updateNextField),String::from(token));
                    updateNextField = "";
                    step = tokens::Step::validateUpdateCommaOrWhere;
                }
                tokens::Step::validateUpdateCommaOrWhere => {
                    if String::from("WHERE").eq_ignore_ascii_case(&token[..]){
                        step = tokens::Step::whereField;
                        continue;
                    }
                    else if String::from(",").eq_ignore_ascii_case(&token[..]){
                        step = tokens::Step::updateField;
                        continue;
                    }
                    else {
                        //TODO: throw proper error code
                        return (q, ast::ErrorCode::ParseError(
                            String::from("at UPDATE: expected , or WHERE")
                        )
                        );
                    }
                }
                tokens::Step::stepWhere  => {
                    if !String::from("WHERE").eq_ignore_ascii_case(&token[..]){
                        //TODO: throw proper error code
                        return (q, ast::ErrorCode::ParseError(
                            String::from("expected WHERE")
                        )
                        );
                    }
                    step = tokens::Step::whereField;
                }
                tokens::Step::whereField =>{
                    if !tokenizer.is_identifier(token){
                        //TODO: throw proper error code
                        return (q, ast::ErrorCode::ParseError(
                            String::from("at WHERE: expected field")
                        )
                        );
                    }
                    let mut c = ast::Condition::new();
                    c.operand1 = String::from(&token[..]);
                    q.conditions.push(c);
                    step = tokens::Step::whereOperator;
                }
                tokens::Step::whereOperator =>{
                    let mut c = q.conditions.pop().unwrap();
                    if String::from("=").eq_ignore_ascii_case(&token[..]) {
                        c.operator = ast::Operator::eq;
                    }
                    else if String::from(">").eq_ignore_ascii_case(&token[..]){
                        c.operator = ast::Operator::gt;
                    }
                    else if String::from(">=").eq_ignore_ascii_case(&token[..]){
                        c.operator = ast::Operator::gte;
                    }
                    else if String::from("<").eq_ignore_ascii_case(&token[..]){
                        c.operator = ast::Operator::lt;
                    }
                    else if String::from("<=").eq_ignore_ascii_case(&token[..]){
                        c.operator = ast::Operator::lte;
                    }
                    else if String::from("!=").eq_ignore_ascii_case(&token[..]){
                        c.operator = ast::Operator::ne;
                    }
                    else {
                        //TODO: throw proper error code
                        return (q, ast::ErrorCode::ParseError(
                            String::from("at WHERE: unknown operator")
                        ));
                    }
                    q.conditions.push(c);
                    step = tokens::Step::whereValue;
                }
                tokens::Step::whereValue  => {
                    if token.len() == 0 {
                        //TODO: throw proper error code
                        return (q, ast::ErrorCode::ParseError(
                            String::from("at WHERE: expected quoted value")
                        )
                        );
                    }
                    let mut c = q.conditions.pop().unwrap();
                    c.operand2 = String::from(&token[..]);
                    c.operand_2_is_field = false;
                    q.conditions.push(c);
                    step = tokens::Step::whereAnd;
                }
                tokens::Step::whereAnd =>{
                    if  ! String::from("AND").eq_ignore_ascii_case(token){
                        //TODO: throw proper error code
                        return (q, ast::ErrorCode::ParseError(
                            String::from("expected AND")
                        ));
                    
                    }
                    step = tokens::Step::whereField;
                }
                tokens::Step::insertFieldsOpeningParens =>{
                    if token.len() != 1 && String::from("(").eq_ignore_ascii_case(&token[..]) {
                        //TODO: throw proper error code
                        return (q, ast::ErrorCode::ParseError(
                            String::from("at INSERt INTO: expected (")

                        ));
                    }
                    step = tokens::Step::insertFields;
                }
                tokens::Step::insertFields =>{
                    if ! tokenizer.is_identifier(token){
                        //TODO: throw proper error code
                        return (q, ast::ErrorCode::ParseError(
                            String::from("at INSERT INTO: expected at least one field to insert")
                        )
                        );
                    }
                    q.fields.push(String::from(token));
                    step = tokens::Step::insertFieldsCommaOrClosingParens;
                }
                tokens::Step::insertFieldsCommaOrClosingParens =>{
                    if token != "," && token != ")" {
                        //TODO: throw proper error code
                        return (q, ast::ErrorCode::ParseError(
                            String::from("at INSERT INTO: expected comma or closing parens")
                        )
                        );
                    }
                    if token == ","{
                        step = tokens::Step::insertFields;
                        continue;
                    }
                    step = tokens::Step::insertValuesRWord;
                }
                tokens::Step::insertValuesRWord =>{
                    if ! String::from("VALUES").eq_ignore_ascii_case(&token[..]){
                        //TODO: throw proper error code
                        return (q, ast::ErrorCode::ParseError(
                            String::from("at INSERT INTO: expected 'VALUES' ")
                        ));
                    }
                    step = tokens::Step::insertValuesOpeningParens;
                }
                tokens::Step::insertValuesOpeningParens =>{
                    if token != "(" {
                        //TODO: throw proper error code
                        return (q, ast::ErrorCode::ParseError(
                            String::from("at INSERT INTO: expected opening parens")
                        ));
                    }
                    q.insert_fields.push(Vec::new());
                    step = tokens::Step::insertValues;
                }
                tokens::Step::insertValues =>{
                    if token.len() == 0 {
                        //TODO: throw proper error code
                        return (q, ast::ErrorCode::ParseError(
                            String::from("at INSERT INTO: expected quoted value")
                        ));
                    }
                    let mut lval = q.insert_fields.pop().unwrap();
                    lval.push(String::from(token));
                    q.insert_fields.push(lval);
                    step = tokens::Step::insertValuesCommaOrClosingParens;
                }
                tokens::Step::insertValuesCommaOrClosingParens =>{
                    if token != "," && token != ")" {
                         //TODO: throw proper error code
                         return (q, ast::ErrorCode::ParseError(
                             String::from("at INSERT INTO: expected comma or closing parens")
                         ));
                    }
                    if token == "," {
                        step = tokens::Step::insertValues;
                        continue;
                    }
                    let last = q.insert_fields.pop().unwrap();
                    if last.len() < q.fields.len(){
                        //TODO: throw proper error code
                        return (q, ast::ErrorCode::ParseError(
                            String::from("at INSERT INTO: value count does not match field count")
                        )
                        );                        
                    }
                    step = tokens::Step::insertValuesCommaBeforeOpeningParens;
                }
                tokens::Step::insertValuesCommaBeforeOpeningParens =>{
                    if token != "," {
                        //TODO: throw proper error code
                        return (q, ast::ErrorCode::ParseError(
                            String::from("at INSERT INTO: expected comma")
                    )
                        );   
                    }
                    step = tokens::Step::insertValuesOpeningParens;
                }
                
            }
            
        }
        return (q, ast::ErrorCode::None)
        
    }
    
    
}

File: new_go_parser/parser.go
package main
import (
	"fmt"
	"regexp"
	"strings"
    "query"
    "Token"

	//"github.com/marianogappa/sqlparser/query"
)
type Parser struct{
    Tokens []string
    i int
    step            step
	query           query.Query
	err             error
	nextUpdateField string
    sql string
    sql_index int
}

func (p *Parser) parse() (query.Query, error) {
    //fmt.println
	q, err := p.doParse()
	p.err = err
	if p.err == nil {
		p.err = p.validate()
	}
	p.logError()
	return q, p.err
}

func (arr *Parser) peek() (string) {
    //&Parser{Tokens,0, stepType, query.Query{}, nil, "",strings.TrimSpace(sql)}
    //fmt.Println("peeked")
    //fmt.Println(arr.Tokens[arr.i])
    //fmt.Println(arr.i)
    return arr.Tokens[arr.i]
}

func (arr *Parser) pop() (string) {
    
    popped:= arr.Tokens[arr.i]
    arr.i=arr.i+1
    arr.sql_index= arr.sql_index+len(popped)
    //fmt.Println("popped")
    //fmt.Println(popped)
    //fmt.Println(arr.i)
    return popped
}
//func (arr *Parser ) copy_tokenarr( token_arr []string ) {
    
  //  for i :=0 ; i < len(token_arr); i++ {
        
    //    arr.Tokens=append(arr.Tokens,token_arr[i])
    //}
    //fmt.Printf("%v", arr.Tokens)
//}

var reservedWords = []string{
	"(", ")", ">=", "<=", "!=", ",", "=", ">", "<", "SELECT", "INSERT INTO", "VALUES", "UPDATE", "DELETE FROM",
	"WHERE", "FROM", "SET",
}

type step int

const (
	stepType step = iota
	stepSelectField
	stepSelectFrom
	stepSelectComma
	stepSelectFromTable
	stepInsertTable
	stepInsertFieldsOpeningParens
	stepInsertFields
	stepInsertFieldsCommaOrClosingParens
	stepInsertValuesOpeningParens
	stepInsertValuesRWord
	stepInsertValues
	stepInsertValuesCommaOrClosingParens
	stepInsertValuesCommaBeforeOpeningParens
	stepUpdateTable
	stepUpdateSet
	stepUpdateField
	stepUpdateEquals
	stepUpdateValue
	stepUpdateComma
	stepDeleteFromTable
	stepWhere
	stepWhereField
	stepWhereOperator
	stepWhereValue
	stepWhereAnd
)


func (p *Parser) doParse() (query.Query, error) {
    //fmt.Println(p.i)
    //fmt.Println(p.Tokens)
    //fmt.Println(p.sql)
	for {
		if p.i >= len(p.Tokens) {
			return p.query, p.err
           // if p.sql_index >= len(p.sql) {
             //   return p.query, p.err
		}
        //fmt.Println(strings.ToUpper(p.peek()))
		switch p.step {
		case stepType:
			switch strings.ToUpper(p.peek()) {
			case "SELECT":
				p.query.Type = query.Select
                //fmt.Println(p.i)
				p.pop()
                //fmt.Println(p.i)
				p.step = stepSelectField
			case "INSERT INTO":
				p.query.Type = query.Insert
				p.pop()
				p.step = stepInsertTable
			case "UPDATE":
				p.query.Type = query.Update
				p.query.Updates = map[string]string{}
				p.pop()
				p.step = stepUpdateTable
			case "DELETE FROM":
				p.query.Type = query.Delete
				p.pop()
				p.step = stepDeleteFromTable
			default:
				return p.query, fmt.Errorf("invalid query type")
			}
		case stepSelectField:
			identifier := p.peek()
			if !isIdentifierOrAsterisk(identifier) {
				return p.query, fmt.Errorf("at SELECT: expected field to SELECT")
			}
			p.query.Fields = append(p.query.Fields, identifier)
			p.pop()
			maybeFrom := p.peek()
			if strings.ToUpper(maybeFrom) == "FROM" {
				p.step = stepSelectFrom
				continue
			}
			p.step = stepSelectComma
		case stepSelectComma:
			commaRWord := p.peek()
			if commaRWord != "," {
				return p.query, fmt.Errorf("at SELECT: expected comma or FROM")
			}
			p.pop()
			p.step = stepSelectField
		case stepSelectFrom:
			fromRWord := p.peek()
			if strings.ToUpper(fromRWord) != "FROM" {
				return p.query, fmt.Errorf("at SELECT: expected FROM")
			}
			p.pop()
			p.step = stepSelectFromTable
		case stepSelectFromTable:
            //fmt.Println("tttttt")
            //fmt.Println(p.i)
            //fmt.Println(p.peek())
			tableName := p.peek()
			if len(tableName) == 0 {
				return p.query, fmt.Errorf("at SELECT: expected quoted table name")
			}
			p.query.TableName = tableName
			p.pop()
			p.step = stepWhere
		case stepInsertTable:
			tableName := p.peek()
			if len(tableName) == 0 {
				return p.query, fmt.Errorf("at INSERT INTO: expected quoted table name")
			}
			p.query.TableName = tableName
			p.pop()
			p.step = stepInsertFieldsOpeningParens
		case stepDeleteFromTable:
			tableName := p.peek()
			if len(tableName) == 0 {
				return p.query, fmt.Errorf("at DELETE FROM: expected quoted table name")
			}
			p.query.TableName = tableName
			p.pop()
			p.step = stepWhere
		case stepUpdateTable:
			tableName := p.peek()
			if len(tableName) == 0 {
				return p.query, fmt.Errorf("at UPDATE: expected quoted table name")
			}
			p.query.TableName = tableName
			p.pop()
			p.step = stepUpdateSet
		case stepUpdateSet:
			setRWord := p.peek()
			if setRWord != "SET" {
				return p.query, fmt.Errorf("at UPDATE: expected 'SET'")
			}
			p.pop()
			p.step = stepUpdateField
		case stepUpdateField:
			identifier := p.peek()
			if !isIdentifier(identifier) {
				return p.query, fmt.Errorf("at UPDATE: expected at least one field to update")
			}
			p.nextUpdateField = identifier
			p.pop()
			p.step = stepUpdateEquals
		case stepUpdateEquals:
			equalsRWord := p.peek()
			if equalsRWord != "=" {
				return p.query, fmt.Errorf("at UPDATE: expected '='")
			}
			p.pop()
			p.step = stepUpdateValue
		case stepUpdateValue:
            //fmt.Println(quotedValue)
            //quotedValue, ln := peekQuoted(p.sql,p.sql_index)
            quotedValue := p.peek()
            ln :=len(quotedValue)
            //fmt.Println(quotedValue)
			if ln == 0 {
				return p.query, fmt.Errorf("at UPDATE: expected quoted value")
			}
			p.query.Updates[p.nextUpdateField] = quotedValue
			p.nextUpdateField = ""
			p.pop()
			maybeWhere := p.peek()
			if strings.ToUpper(maybeWhere) == "WHERE" {
				p.step = stepWhere
				continue
			}
			p.step = stepUpdateComma
		case stepUpdateComma:
			commaRWord := p.peek()
			if commaRWord != "," {
				return p.query, fmt.Errorf("at UPDATE: expected ','")
			}
			p.pop()
			p.step = stepUpdateField
		case stepWhere:
			whereRWord := p.peek()
			if strings.ToUpper(whereRWord) != "WHERE" {
				return p.query, fmt.Errorf("expected WHERE")
			}
			p.pop()
			p.step = stepWhereField
		case stepWhereField:
			identifier := p.peek()
			if !isIdentifier(identifier) {
				return p.query, fmt.Errorf("at WHERE: expected field")
			}
			p.query.Conditions = append(p.query.Conditions, query.Condition{Operand1: identifier, Operand1IsField: true})
			p.pop()
			p.step = stepWhereOperator
		case stepWhereOperator:
			operator := p.peek()
			currentCondition := p.query.Conditions[len(p.query.Conditions)-1]
			switch operator {
			case "=":
				currentCondition.Operator = query.Eq
			case ">":
				currentCondition.Operator = query.Gt
			case ">=":
				currentCondition.Operator = query.Gte
			case "<":
				currentCondition.Operator = query.Lt
			case "<=":
				currentCondition.Operator = query.Lte
			case "!=":
				currentCondition.Operator = query.Ne
			default:
				return p.query, fmt.Errorf("at WHERE: unknown operator")
			}
			p.query.Conditions[len(p.query.Conditions)-1] = currentCondition
			p.pop()
			p.step = stepWhereValue
		case stepWhereValue:
			//quotedValue, ln := peekQuoted(p.sql,p.sql_index)
             quotedValue := p.peek()
            ln :=len(quotedValue)
            //fmt.Println(ln)
            //fmt.Println(p.sql)
            
            //fmt.Println(p.sql_index)
			if ln == 0 {
				return p.query, fmt.Errorf("at WHERE: expected quoted value")
			}
			currentCondition := p.query.Conditions[len(p.query.Conditions)-1]
			currentCondition.Operand2 = quotedValue
			currentCondition.Operand2IsField = false
			p.query.Conditions[len(p.query.Conditions)-1] = currentCondition
			p.pop()
			p.step = stepWhereAnd
		case stepWhereAnd:
			andRWord := p.peek()
			if strings.ToUpper(andRWord) != "AND" {
				return p.query, fmt.Errorf("expected AND")
			}
			p.pop()
			p.step = stepWhereField
		case stepInsertFieldsOpeningParens:
			openingParens := p.peek()
			if len(openingParens) != 1 || openingParens != "(" {
				return p.query, fmt.Errorf("at INSERT INTO: expected opening parens")
			}
			p.pop()
			p.step = stepInsertFields
		case stepInsertFields:
			identifier := p.peek()
			if !isIdentifier(identifier) {
				return p.query, fmt.Errorf("at INSERT INTO: expected at least one field to insert")
			}
			p.query.Fields = append(p.query.Fields, identifier)
			p.pop()
			p.step = stepInsertFieldsCommaOrClosingParens
		case stepInsertFieldsCommaOrClosingParens:
			commaOrClosingParens := p.peek()
			if commaOrClosingParens != "," && commaOrClosingParens != ")" {
				return p.query, fmt.Errorf("at INSERT INTO: expected comma or closing parens")
			}
			p.pop()
			if commaOrClosingParens == "," {
				p.step = stepInsertFields
				continue
			}
			p.step = stepInsertValuesRWord
		case stepInsertValuesRWord:
			valuesRWord := p.peek()
			if strings.ToUpper(valuesRWord) != "VALUES" {
				return p.query, fmt.Errorf("at INSERT INTO: expected 'VALUES'")
			}
			p.pop()
			p.step = stepInsertValuesOpeningParens
		case stepInsertValuesOpeningParens:
			openingParens := p.peek()
			if openingParens != "(" {
				return p.query, fmt.Errorf("at INSERT INTO: expected opening parens")
			}
			p.query.Inserts = append(p.query.Inserts, []string{})
			p.pop()
			p.step = stepInsertValues
		case stepInsertValues:
			//quotedValue, ln := peekQuoted(p.sql,p.sql_index)
             quotedValue := p.peek()
            ln :=len(quotedValue)
			if ln == 0 {
				return p.query, fmt.Errorf("at INSERT INTO: expected quoted value")
			}
			p.query.Inserts[len(p.query.Inserts)-1] = append(p.query.Inserts[len(p.query.Inserts)-1], quotedValue)
			p.pop()
			p.step = stepInsertValuesCommaOrClosingParens
		case stepInsertValuesCommaOrClosingParens:
			commaOrClosingParens := p.peek()
			if commaOrClosingParens != "," && commaOrClosingParens != ")" {
				return p.query, fmt.Errorf("at INSERT INTO: expected comma or closing parens")
			}
			p.pop()
			if commaOrClosingParens == "," {
				p.step = stepInsertValues
				continue
			}
			currentInsertRow := p.query.Inserts[len(p.query.Inserts)-1]
			if len(currentInsertRow) < len(p.query.Fields) {
				return p.query, fmt.Errorf("at INSERT INTO: value count doesn't match field count")
			}
			p.step = stepInsertValuesCommaBeforeOpeningParens
		case stepInsertValuesCommaBeforeOpeningParens:
			commaRWord := p.peek()
			if strings.ToUpper(commaRWord) != "," {
				return p.query, fmt.Errorf("at INSERT INTO: expected comma")
			}
			p.pop()
			p.step = stepInsertValuesOpeningParens
		}
	}
}

func (p *Parser) validate() error {
	if len(p.query.Conditions) == 0 && p.step == stepWhereField {
		return fmt.Errorf("at WHERE: empty WHERE clause")
	}
	if p.query.Type == query.UnknownType {
		return fmt.Errorf("query type cannot be empty")
	}
	if p.query.TableName == "" {
		return fmt.Errorf("table name cannot be empty")
	}
	if len(p.query.Conditions) == 0 && (p.query.Type == query.Update || p.query.Type == query.Delete) {
		return fmt.Errorf("at WHERE: WHERE clause is mandatory for UPDATE & DELETE")
	}
	for _, c := range p.query.Conditions {
		if c.Operator == query.UnknownOperator {
			return fmt.Errorf("at WHERE: condition without operator")
		}
		if c.Operand1 == "" && c.Operand1IsField {
			return fmt.Errorf("at WHERE: condition with empty left side operand")
		}
		if c.Operand2 == "" && c.Operand2IsField {
			return fmt.Errorf("at WHERE: condition with empty right side operand")
		}
	}
	if p.query.Type == query.Insert && len(p.query.Inserts) == 0 {
		return fmt.Errorf("at INSERT INTO: need at least one row to insert")
	}
	if p.query.Type == query.Insert {
		for _, i := range p.query.Inserts {
			if len(i) != len(p.query.Fields) {
				return fmt.Errorf("at INSERT INTO: value count doesn't match field count")
			}
		}
	}
	return nil
}


func (p *Parser) logError() {
	if p.err == nil {
		return
	}
	fmt.Printf("%v", p.Tokens)
	fmt.Println(strings.Repeat(" ", p.i) + "^")
	fmt.Println(p.err)
}

func isIdentifier(s string) bool {
	for _, rw := range reservedWords {
		if strings.ToUpper(s) == rw {
			return false
		}
	}
	matched, _ := regexp.MatchString("[a-zA-Z_][a-zA-Z_0-9]*", s)
	return matched
}

func isIdentifierOrAsterisk(s string) bool {
	return isIdentifier(s) || s == "*"
}

//func peekQuoted(s string, i int) (string, int) {
//	if len(s) < i || s[i] != '\'' {
//		return "", 0
//	}
//	for j := i + 1; j < len(s); j++ {
//		if s[j] == '\'' && s[j-1] != '\\' {
//			return s[i+1 : j], len(s[i+1:j]) + 2 // +2 for the two quotes
//		}
//	}
//	return "", 0
//}
func main() {
    var sql string ="SELECT a, c, d FROM 'b' WHERE a > '1'"
    //fmt.Println(strings.TrimSpace(sql))
//parser{0, strings.TrimSpace(sql), stepType, query.Query{}, nil, ""}
    // var token_arr []string
    var token_arr []string = (&Token.Tokenizer{0, strings.TrimSpace(sql), nil}).Pop()
   // func (p *Tokenizer) peekQuotedStringWithLength() (string, int) {
   // fmt.Printf("%v", token_arr)
    //var Tokens []string
    //(&Parser{Tokens,0, stepType, query.Query{}, nil, "",strings.TrimSpace(sql)}).copy_tokenarr(token_arr)
    query, err := (&Parser{token_arr,0, stepType, query.Query{}, nil, "",strings.TrimSpace(sql),0}).parse()
	//if err != nil {
	//	log.Fatal(err)
	//}
	fmt.Printf("%+#v", query)
    //fmt.println(err)
    fmt.Println(err)
   
     
}

